enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,284 @@
+This program is executed in the NIC’s device driver, right after the interrupt processing and before the network stack. 
+The goal of XDP is to offer comparable performance to kernel bypass solutions while working with the existing kernel networking stack.
+A new cluster-scoped object `IngressNodeFirewall` will be used to configure node ingress firewall match rules and per rule statistics will be avaliable.
+Admin can provision multiple instances of this object while ensuring certain execution order.
+For the best performance XDP programs need to attach to interfaces using the following models:
+- Native XDP \
+  The XDP program is loaded by the network card driver as part of its initial receive path. This also requires support from the network card driver.
+- Offloaded XDP \
+  The XDP program loads directly on the NIC, and executes without using the CPU. This also requires support from the network interface device.
+
+## Motivation
+
+Customers would like to secure OCP cluster from external attacks.
+
+### Goals
+
+- Flexible match filter rules for customers to use with a CRD validation to detect and reject invalid configuration.
+- Provide a mechanism for customers to select which nodes via `nodeSelector`.
+- Provide a mechanism to define attachment interfaces for the node ingress firewall objects.  
+- Ingress node firewall is deployed on a fully operational OCP cluster (i.e. Ingress node firewall is deployed as a day 2 operation and is not enabled by default at the cluster install time).
+- fail safe in case of user provisioned invalid rules which can lockup cluster operation.
+
+### Non-Goals
+- Stateful support
+- Flood Attacks
+
+## Proposal
+
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will deploy daemonSet controller on all the selected nodes by applying `IngressNodeFirewallConfig` object. 
+- customers will prepare `IngressNodeFirewall` objects with all required firewall rules and interfaces the ebpf program will be attached to.
+- IngressNodeFW controller will process `IngressNodeFirewall` objects, verification webhook will make sure configuration doesn't contradict with fail safe rules.
+- Ingress node firewall daemon will watch ingress node firewall objects and create `nodeState` object which contains all rules objects that is applied to specific node then load eBPF map table and attach eBPF kernel hook loaded as `BPF_PROG_TYPE_XDP` to the list of interfaces configured in the object.
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
+
+### Workflow Description
+
+- the following diagram shows how node ingress firewall configuration is handled
+
+![Ingress Firewall configuration processing](./ingress-nodefw-config-flow.png)
+
+- the following diagram shows eBPF how ingress firewall kernel hook will process packets and apply firewall rules
+
+![eBPF workflow](./ebpf-packet-flow.png)
```

Comment:
ugggh... "everything is accepted by default, except if you create an IngressNodeFirewall in which case everything is rejected by default" is a terrible semantic. This is how NetworkPolicy works and it's literally the worst problem with NetworkPolicy.

IngressNodeFirewall allows the admin to explicitly say "and reject everything else". If that's the semantic they want, then they should say that explicitly.

Reply:
In particular, IngressNodeFirewall is designed so that you can have multiple INF objects. But the "if there are any INFs then default deny" rule means that INFs don't compose correctly. Especially:
1. You can't create a new IngressNetworkFirewall object if you don't know whether or not there are other existing IngressNetworkFirewall objects. Creating an object that says "Deny access to port 999" could end up meaning either "deny access to port 999" or "deny access to literally everything except the failsafe stuff".
2. You can't work around that by doing "high-priority deny access to port 999, else low-priority allow access to everything", because the cluster might otherwise be depending on the default-deny-everything rule and then you'd break that.

Reply:
we felt we have to move from what we have today were we will let everything to the cluster , to select if we need to drop very specific patterns while keep the default to allow and that is indeed will be very clear in our doc. and ofcourse up to improvement as we get more feedback

Reply:
Regardless of what we want the default behavior to be or what customers want the default behavior to be, having the default behavior _change_ when you create the first IngressNodeFirewall is bad and makes the feature confusing and difficult to work with.

If someone wants "everything else is blocked", then they can just create a rule with a very high `order` matching all traffic and denying it. We can even provide an example of this in the documentation for them to comment.

Reply:
so is this what u recommend we doc if someone wanted to default to drop set the a rule with early order to drop all ? at least if he encounter critical service drops he will need to fine tune his rule ?

Reply:
I don't care where the "else drop all" comes from, as long as it is _explicit_ somewhere. The user could have to specify it in details themselves, or there could be a special rule type for it, or there could be a second CRD where you set a flag saying "I want default deny behavior". But it should not just happen automatically any time the number of IngressNodeFirewalls changes from 0 to 1.

Reply:
I prefer to leave it to customer to explicitly specify it for themselves. so they are in control of the behavior.